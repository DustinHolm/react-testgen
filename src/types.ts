export interface ImportBlock {
    sourceFile: string,
    isInternal: boolean,
    imports: Import[]
}

export interface Import {
    isDefault: boolean,
    name: string,
    element?: ExportableElement
}

export interface Export {
    name: string,
    isDefault: boolean,
    element?: ExportableElement
}

export enum ExportType {
    Class,
    Function,
    Interface,
    Variable
}

export interface ExportableElement {
    type: ExportType,
    returnsJSX: boolean
}

export interface ClassElement extends ExportableElement {
    methods: MethodElement[],
    constructors: Constructor[]
}

export interface MethodElement extends FunctionElement {
    name: string
}

export interface FunctionElement extends ExportableElement {
    returnType: string,
    parameters: Attribute[]
}

export interface InterfaceElement extends ExportableElement {
    attributes: Attribute[]
}

export interface VariableElement extends ExportableElement {
    varType: string,
    varValue: string
}

export interface Attribute {
    name: string,
    type: string
}

export interface Constructor {
    parameters: Attribute[]
}