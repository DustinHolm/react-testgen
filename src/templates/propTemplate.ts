import { Attribute } from "../types"
import { getReturnForType, toPascalCase } from "./common"

export const createProp = (prop: Attribute): string => {
    const name = "mockProp" + toPascalCase(prop.name)
    const pascalName = toPascalCase(name)

    return `\
let ${name}: ${prop.type}
function reset${pascalName}() {
    ${name} = ${getReturnForType(prop.type)}
}
function given${pascalName}(given: ${prop.type}) {
    ${name} = given
}
`
}