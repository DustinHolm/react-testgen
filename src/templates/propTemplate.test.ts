import { Attribute } from "../types"
import { createProp } from "./propTemplate"

describe("propTemplate Tests", () => {
    beforeEach(() => {
        resetMocks()
    })

    test("With attribute", () => {
        const returnedString = createProp({ name: "testName", type: "testType" })
        expect(returnedString).toMatchSnapshot()
    })

    test("getReturnForType is called correctly", () => {
        const attribute: Attribute = { name: "testName", type: "testType" }
        createProp(attribute)
        expect(mockGetReturnForType).toHaveBeenCalledWith([attribute.type])
    })
})

function resetMocks() {
    resetMockGetReturnForType()
}

let mockGetReturnForType: jest.Mock
function resetMockGetReturnForType() {
    mockGetReturnForType = jest.fn(() => "mockedReturn")
}
jest.mock("./common", () => ({
    ...jest.requireActual("./common"),
    getReturnForType: (...args: any[]) => mockGetReturnForType(args)
}))
