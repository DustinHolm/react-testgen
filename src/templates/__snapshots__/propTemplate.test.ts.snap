// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propTemplate Tests With attribute 1`] = `
"let mockPropTestName: testType
function resetMockPropTestName() {
    mockPropTestName = mockedReturn
}
function givenMockPropTestName(given: testType) {
    mockPropTestName = given
}
"
`;
